universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,199 @@
+#include "ValueRefs.h"
+
+#include <functional>
+#include <iomanip>
+#include <iterator>
+#include <unordered_map>
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/numeric.hpp>
+#include "Enums.h"
+#include "Field.h"
+#include "Fighter.h"
+#include "Fleet.h"
+#include "Pathfinder.h"
+#include "Planet.h"
+#include "ShipDesign.h"
+#include "ShipHull.h"
+#include "ShipPart.h"
+#include "Ship.h"
+#include "Species.h"
+#include "System.h"
+#include "Tech.h"
+#include "UniverseObjectVisitors.h"
+#include "UniverseObject.h"
+#include "Universe.h"
+#include "../Empire/EmpireManager.h"
+#include "../Empire/Empire.h"
+#include "../Empire/Supply.h"
+#include "../util/GameRules.h"
+#include "../util/Logger.h"
+#include "../util/MultiplayerCommon.h"
+#include "../util/Random.h"
+
+NamedValueRefManager* NamedValueRefManager::s_instance = nullptr;
+
+NamedValueRefManager::NamedValueRefManager() {
+    if (s_instance)
+        throw std::runtime_error("Attempted to create more than one NamedValueRefManager.");
+
+    // Only update the global pointer on sucessful construction.
+    ErrorLogger() << "NamedValueRefManager::NameValueRefManager constructs singleton " << this;
+    s_instance = this;
+}
+
+NamedValueRefManager::~NamedValueRefManager() {
+    ErrorLogger() << "NamedValueRefManager::~NameValueRefManager destruct " << this ; // FIXME
+}
+
+void NamedValueRefManager::SetNamedValueRefs(Pending::Pending<NamedValueRefManager::container_type>&& future)
+{ m_pending_refs = std::move(future); }
+
+
+void NamedValueRefManager::CheckPendingRefs() const {
+    ErrorLogger() << "Number of registered ValueRefs befor swap: " << m_value_refs.size() << " in " << this;
+    if (auto tt = Pending::WaitForPending(m_pending_refs)) {
+      std::swap(*tt, const_cast<NamedValueRefManager::container_type&>(m_value_refs));      
+	for (auto& k_v : *tt) {
+	  ErrorLogger() << "NamedValueRefManager::CheckPendingRefs restore swapped away " << k_v.first;
+	  const_cast<NamedValueRefManager::container_type&>(m_value_refs).emplace(k_v.first, std::move(k_v.second));
+	}
+	(*tt).clear();
+    }
+    ErrorLogger() << "Number of registered ValueRefs after swap: " << m_value_refs.size() << " in " << this;
+
+}
+
+template <typename T>
+ValueRef::ValueRef<T>* NamedValueRefManager::GetValueRef(const std::string& name) /*const*/ {
+    InfoLogger() << "NamedValueRefManager::GetValueRef<T> look for registered valueref for \"" << name << '"';
+    auto* vref = GetAnyValueRef(name);
+    if (vref) {
+        return dynamic_cast<ValueRef::ValueRef<T>*>(vref);
```

Comment:
probably would be good to avoid using `dynamic_cast` on every call use `GetValueRef` if possible...

Reply:
that code **is** the definition of `GetValueRef`. I do not think there is a way not to use dynamic_cast with the `m_value_refs` defined as `map<string, unique_ptr<ValueRefBase>>` or is there? 

Not sure what options are there

1. split the map into multiple maps (maybe 'map<string, unique_ptr<ValueRef<double>>' and int and one for everything else and query those maps one after the other (with double being the most important) if the necessary type is unknown
2. restructure the ValueRefs so they have a unified structure/data layout
3. use some magic multi type map someone else wrote 
4. do something so the dynamic_cast does not matter - e.g. cache the valueref in the NamedRef (weak ptr or sth) so the dynamic cast has only to be done once; for UI/string representation the type does not matter - the value text is already polymorphic
5. use some variant stuff
6. weaken type use in NamedRef - use some `<template ExpectedType> ExpectedType ValueRefBase::EvalOrBust()`   which returns a default value (or try to convert) for other types (e.g. `Variable<int>` would return the string representation of the Eval result for `EvalOrBust<string>`)
7. something simpler i did not think about

suggestions very appreciated

Reply:
I suggested separate maps below...

don't need multiple queries if you know the type `T`

Reply:
alright, 1.) separate maps its gonna be